theano/gpuarray/type.py

```diff
@@ -250,6 +257,12 @@ def filter(self, data, strict=False, allow_downcast=None):
             if b and shp[i] != 1:
                 raise TypeError("Non-unit value on shape on a broadcastable"
                                 " dimension.", shp, self.broadcastable)
+        if not isinstance(data, gpuarray.GpuArray):
+            if old_data and old_data.shape == data.shape:
+                old_data.write(data)
```

Comment:
Both need to be contiguous for it to work.  However if data is not contiguous it will be copied (on the CPU) to a contiguous buffer.  If old_data is not contiguous it will error out.